### PR TITLE
Vcluster/2412 bulk edit contracts

### PR DESCRIPTION
**Pull request by ckr123**

### CODE REVIEW

File: domains/certificates/Query.API/API/ContractService/ContractServiceImpl.cs [hunk 1]
```diff
@@ -115,41 +115,52 @@ private static MeteringPointType Map(MeterType type)
...
-        if (contract.MeteringPointOwner != meteringPointOwner)
+        foreach (var (id, newEndDate) in contracts)
         {
-            return new MeteringPointOwnerNoMatch();
-        }
-
-        if (newEndDate.HasValue && newEndDate <= contract.StartDate)
-        {
-            return new EndDateBeforeStartDate(contract.StartDate, newEndDate.Value);
+            var contract = await unitOfWork.CertificateIssuingContractRepo.GetById(id, cancellationToken);
```
Comment (TopSwagCode) [inline domains/certificates/Query.API/API/ContractService/ContractServiceImpl.cs:124]: We should not GetById in a for loop. We should move this up and fetch all Certificates in 1 database call.

**Comment by TopSwagCode** [comment]
We need to bundle database calls. Otherwise I don't see the point of this issue. Then the frontend might as well call several times, if the logic is the same anyways. Might just be me :)

File: domains/certificates/Query.API/API/ContractService/ContractServiceImpl.cs [hunk 2]
```diff
@@ -115,41 +115,52 @@ private static MeteringPointType Map(MeterType type)
...
+            if (contract.MeteringPointOwner != meteringPointOwner)
+            {
+                return new MeteringPointOwnerNoMatch();
+            }
+
+            if (newEndDate != null && newEndDate.ToDateTimeOffset() <= contract.StartDate)
+            {
+                return new EndDateBeforeStartDate(contract.StartDate, newEndDate.ToDateTimeOffset());
+            }
+            var issuingContracts = await unitOfWork.CertificateIssuingContractRepo.GetByGsrn(contract.GSRN, cancellationToken);
```
[the remark this answers is not in the document]
Comment (ckr123) [reply domains/certificates/Query.API/API/ContractService/ContractServiceImpl.cs]: Bundled.
We are still hitting the activitylog hard - but i dont think it is going to be a problem, if it is, i will create i task to bump the nuget

**Comment by TopSwagCode** [comment]
Convert to PR and I will approve :D. Also remove merge conflict :)

**Review by TopSwagCode — APPROVED** [review]

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs [hunk 3]
```diff
@@ -45,21 +48,70 @@ public class ContractsController : ControllerBase
...
-        var startDate = DateTimeOffset.FromUnixTimeSeconds(createContract.StartDate);
-        DateTimeOffset? endDate = createContract.EndDate.HasValue ? DateTimeOffset.FromUnixTimeSeconds(createContract.EndDate.Value) : null;
+        var startDate = UnixTimestamp.Create(createContract.StartDate);
+        UnixTimestamp? endDate = createContract.EndDate.HasValue
+            ? UnixTimestamp.Create(createContract.EndDate.Value)
+            : null;
 
         var result = await service.Create(
-            createContract.GSRN,
+            [(createContract.GSRN, startDate, endDate)],
```
Comment (TopSwagCode) [inline domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs:57]: Why take list as argument if we only call it with 1 object?

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs [hunk 4]
```diff
@@ -45,21 +48,70 @@ public class ContractsController : ControllerBase
...
+    [ApiVersion(ApiVersions.Version20240423)]
+    [Route("api/certificates/contracts")]
+    public async Task<ActionResult> CreateContract(
+        [FromBody] CreateContracts createContracts,
+        [FromServices] IValidator<CreateContract> validator,
+        [FromServices] IContractService service,
+        CancellationToken cancellationToken)
+    {
+        var user = new UserDescriptor(User);
+        var certificateIssuingContracts = new List<(string gsrn, UnixTimestamp startDate, UnixTimestamp? endDate)>();
```
Comment (TopSwagCode) [inline domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs:87]: Couldn't "(string gsrn, UnixTimestamp startDate, UnixTimestamp? endDate)" be a class / record?

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs [hunk 5]
```diff
@@ -130,19 +182,69 @@ public class ContractsController : ControllerBase
             return ValidationProblem(ModelState);
         }
 
-        DateTimeOffset? newEndDate = editContractEndDate.EndDate.HasValue ? DateTimeOffset.FromUnixTimeSeconds(editContractEndDate.EndDate.Value) : null;
+        var newEndDate = editContractEndDate.EndDate.HasValue
+            ? UnixTimestamp.Create(editContractEndDate.EndDate.Value)
+            : null;
+        var contracts = new List<(Guid id, UnixTimestamp? newEndDate)> { (id, newEndDate) };
```
Comment (TopSwagCode) [inline domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs:188]: maybe have a contractDTO or similar?

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs (hunk 4 above)
Comment (ckr123) [reply domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs]: fixed

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs (hunk 5 above)
Comment (ckr123) [reply domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs]: fixed

File: domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs (hunk 3 above)
Comment (ckr123) [reply domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs]: This is endpoint is going to be deprecated in newer versions, so we only need 1 Create method when making a list

**Review by TopSwagCode — APPROVED** [review]

**Pull request merged**